bedwyr/man/bedwyr.1.in
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"  Bedwyr prover                                                           "
.\"  Copyright (C) 2012-2015 Quentin Heath                                   "
.\"                                                                          "
.\"  This program is free software; you can redistribute it and/or modify    "
.\"  it under the terms of the GNU General Public License as published by    "
.\"  the Free Software Foundation; either version 2 of the License, or       "
.\"  (at your option) any later version.                                     "
.\"                                                                          "
.\"  This program is distributed in the hope that it will be useful,         "
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of          "
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           "
.\"  GNU General Public License for more details.                            "
.\"                                                                          "
.\"  You should have received a copy of the GNU General Public License along "
.\"  with this program; if not, write to the Free Software Foundation, Inc., "
.\"  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.             "
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" @configure_input@
.\"
.Dd June 16, 2015
.Os @PACKAGE_NAME@ v@PACKAGE_VERSION@
.Dt BEDWYR 1 URM
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh NAME
.Nm @PACKAGE_TARNAME@ ,
.Nm Ns Nm .byte ,
.Nm Ns Nm .native
.Nd @PACKAGE_TAGLINE@
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh SYNOPSIS
.Nm Op Fl I
.Op Fl t
.Op Fl T
.Op Fl \-filter
.Op Fl d Ar command
.Op Fl e Ar query
.Op Fl \-freezing Ar fp
.Op Fl \-saturation Ar sp
.Op Fl \-version
.Op Fl D
.Op Fl \-colour Ar n
.Op Fl help Ns | Ns Fl \-help
.Op Ar
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh DESCRIPTION
@PACKAGE_NAME@ is a tabled logic programming system for the Level-0/1
fragment of the Linc logic.
.Pp
It is based on
.An Tiu Ns 's
Level-0/1, and
.An Nadathur & Linell Ns 's
LLambda library. For background on the system, see
.Rs
.%T Mixing Finite Success and Finite Failure in an Automated Prover
.%A Alwen Tiu
.%A Gopalan Nadathur
.%A Dale Miller
.%O http://www.lix.polytechnique.fr/Labo/Dale.Miller/papers/eshol05.pdf
.Re
.Pp
Webpage:
.Pa http://slimmer.gforge.inria.fr/bedwyr/ .
.Pp
The wrapper
.Nm
is the preferred invocation.
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh OPTIONS
.Bl -tag -width Ic
.It Fl I
Do not enter interactive mode.
.It Fl t
Run tests (given by
.Ic #assert Ar query Ns Ic \&. ,
see
.Sx META-COMMANDS )
from the definition files given on the command-line.  Use
.Ar n+1
times to test also files that occured under up to
.Ar n
levels of
.Ic #include Qo Ar file.def Qc Ns Ic \&. .
.It Fl T
Run tests in all definition files.
.It Fl \-filter
Use tabling with variables.  If not set, only strict matches are used.
.It Fl d Ar command
Run
.Ar command
(declaration of definition, see
.Sx INPUT FILES ) .
Multiple commands can be given.  They are run after all the input files
are read, and before executing queries.
.It Fl e Ar query
Execute
.Ar query
(see
.Sx TOPLEVEL QUERIES ) .
Multiple queries can be given.  They are executed after all the commands
are run, and before entering the interactive loop.
.It Fl \-freezing Ar fp
Enable backward chaining and set its limit (the so-called
.Ar freezing-point )
to
.Ar fp .
Setting it to 0 (the default value) disables backward chaining,
setting it to \-1 allows for infinite chaining (usually not a smart idea).
The value 1 should be a good compromise.
.It Fl \-saturation Ar sp
Enable forward chaining and set its limit (the so-called
.Ar saturation-point )
to
.Ar sp .
Setting it to 0 (the default value) disables forward chaining,
setting it to \-1 allows infinite chaining and therefore actual saturation
of the table wrt the theorem.
.It Fl \-version
Display version information and a list of compiled features.
.It Fl D
Print debugging information on stderr.  Can be enabled/disabled at
run-time with
.Ic #debug .
.It Fl \-colours Ar n
Force the number of supported colours.
.Ar 256
and
.Ar 8
assume the corresponding number of ANSI colours,
anything else disables the coloured output.
.It Fl help , \-help
Display the list of command-line options.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh EXIT STATUS
.Ex -std
.Bl -tag -width Ic
.It Ar 1
Input error (base syntax error, type error)
.It Ar 2
Command error (wrong declaration, wrong definition)
.It Ar 3
Solver error (LLambda violation, Level 0/1 violation, object misuse)
.It Ar 4
Bedwyr error (failed assertion, meta-command misuse)
.It Ar 5
Undefined error
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh INPUT FILES
A definition file can contain any meta-command (see
.Sx META-COMMANDS ) ,
in addition to any of the following commands.
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Ss Type declaration
.Bl -tag -width Ic
.It Ic Kind Ar list Ic type \-> type.
.Ar nat , string
and
.Ar  prop
are built in, while
.Ar list
and
.Ar option
are predeclared.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Ss Constant declaration
.Bl -tag -width Ic
.It Ic Type Ar :: Ic A \-> list A \-> list A.
.It Ic Type Ar nil Ic list A.
.Ar true
and
.Ar false
are built-ins of type
.Ic prop ,
.Ar 0 , 1 , ...
and any quoted string are runtime-defined constants of type
.Ic nat
and
.Ic string ,
and
.Ar :: ,
.Ar nil ,
.Ar some
and
.Ar none
are predeclared.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Ss Predicate declaration and definition
.Bl -tag -width Ic
.It Ic Define Ar member Ic \&: A \-> list A \-> prop by\
  Ar member X (X :: _) Ic \&;\
  Ar member X (_ :: L) := member X L Ns Ic \&.
.Ic print , println , printstr , fprint , fprintln ,
.Ic fprintstr , fopen_out
and
.Ic fclose_out
are built in,
.Ic member
is predefined.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Ss Theorems
.Bl -tag -width Ic
.It Ic Theorem Ar thm Ic \&: Ar forall Y, body X Y \-> head X Ns Ic \&.
A theorem is a valid Abella theorem with some restrictions on its body
(neither implication nor universal quantification).
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh TOPLEVEL QUERIES
The toplevel input can contain any meta-command (see
.Sx META-COMMANDS ) ,
as well as any formula.
In the latter case, each formula is a query which is solved
.Ns ( Nm
searches for all substitutions of free variables satisfying
.Ar query
and offers to display them one by one).
.Bl -tag -width Ic
.It Ic ?= Ar query Ns Ic \&.
Interactive toplevel.  Queries can span on several lines, but all input
between the
.Ic \&.
and the end of the line is ignored.
.It Nm Fl e Qo Ar query Ns Ic \&. Qc
Non-interactive toplevel.  The string must not be empty, and all input
between the
.Ic \&.
and the end of the line is ignored.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh META-COMMANDS
Meta-commands can be given both in a file and to the toplevel.
.Bl -tag -width Ic
.It Ic #help.
This command should give a starting point.
.It Ic #exit.
And this one, an ending point.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh OUTPUT
Non-interactive output is offerd by
.Ic #save_table Ar pred Qo Ar table.def Qc Ns Ic \&.
or
.Ic #export Qo Ar tables.xml Qc Ns Ic \&. .
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh FILES
.Bl -tag -width Ic
.It Pa +pkgdatadir+/examples/
Sample input files.
Most of them have assertions that can be checked with the command-line option
.Fl t .
.It Pa +pkgdatadir+/contrib/
Text-mode goodies.
.El
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.Sh SEE ALSO
.\".Xr abella 1 ,
.\".Xr tac 1 ,
.Xr teyjus 1


ChangeLog
1.4 (dev)
=========
  Bedwyr-with-round-tables
Changes:
  - port of the build system (bin and odoc) to ocamlbuild
  - zipper-based indexing with get-and-set functionality
New:
  - simple theorems are used to improve tabling
    - forward chaining for filling (#saturation, --saturation)
    - backward chaining for accessing(#freezing, --freezing)
  - *.deb generation handled by the Makefile

1.3 (2012-02-29)
================
  Bedwyr-with-types
Fix:
  - 2 implication bugs : #13229 + timestamp problem (rev 740)
  - no more multiple input with multiple #include
Changes:
  - syntax closer to Abella's :
    - /\ for conjunction instead of ',' or &
    - \/ for disjunction instead of ';'
    - -> instead of =>
    - "forall X Y," instead of "pi X\ pi Y\"
    - "exists X Y," instead of "sigma X\ sigma Y\"
    - "nabla X Y," instead of "nabla X\ nabla Y\"
    - clauses grouped in definition blocks, only one block per predicate,
      no "inductive" and "coinductive" in the same block
    - type annotations for predicates and bound variables
    - Abella's keywords (e.g. Teyjus' keywords) forbidden
  - more names allowed (including names for infix constants)
  - C-style multiline nested comments
  - multiline quoted strings with escaped characters
  - parser less permissive: no more formula-as-term
  - more error output
  - ndcore in a distinct OCaml lib
  - updated userguide and README
New:
  - types
    - objects simply typed (no polymorphism, except for some predefined
      predicates like 'print')
    - input type-checked, declarations mandatory for types (except for nat,
      string and prop), constants and predicates (except for predefined
      predicates)
    - typing commands (#env, #typeof, but no more #type)
    - deprecated 'parse' predicate removed
  - equivariant tabling
  - I/O predefined predicates (print, fprint, etc)
  - undocumented experimental internal non-logical side-effecting
    predicates (_not, _abstract, _abort, etc)
  - table export (#save_table)
  - .deb and .ebuild publicly available

1.2 (2010-04-17)
================
Fix:
 - priority of conjunction and disjunction in the syntax
 - detect potential hole in unification when some left unifications
   make the level-based representation of raising inadequate
New:
 - underscore (_) is allowed for dummy logic variables
 - several new examples
 - pretty printing (indentation and better name generation)
 - slight extension of unification (simple case beyond llambda)

1.1 (2007-02-15)
================
Fix:
  - indexing
Changes:
  - handling of some error related to tabling
New:
  - some examples

1.0 (2006-11-01)
================
  Initial stable release (Bedwyr, the not-so-sound logician)
Features:
  - completed documentation
  - lots of examples
  - usability enhancements: session management, single-line comments
  - assertions for testing
  - new representation of nabla variables avoiding raising
  - major bugfixes in tabling


bedwyr/src/bedwyr.ml
(****************************************************************************)
(* Bedwyr prover                                                            *)
(* Copyright (C) 2005 David Baelde, Alwen Tiu, Axelle Ziegler               *)
(* Copyright (C) 2006 Andrew Gacek                                          *)
(* Copyright (C) 2006,2007,2009,2011 David Baelde                           *)
(* Copyright (C) 2011 Alwen Tiu                                             *)
(* Copyright (C) 2011-2015 Quentin Heath                                    *)
(*                                                                          *)
(* This program is free software; you can redistribute it and/or modify     *)
(* it under the terms of the GNU General Public License as published by     *)
(* the Free Software Foundation; either version 2 of the License, or        *)
(* (at your option) any later version.                                      *)
(*                                                                          *)
(* This program is distributed in the hope that it will be useful,          *)
(* but WITHOUT ANY WARRANTY; without even the implied warranty of           *)
(* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *)
(* GNU General Public License for more details.                             *)
(*                                                                          *)
(* You should have received a copy of the GNU General Public License along  *)
(* with this program; if not, write to the Free Software Foundation, Inc.,  *)
(* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.              *)
(****************************************************************************)

let () =
  let term_width =
    try int_of_string (Sys.getenv "COLUMNS")
    with Failure ("int_of_string") | Not_found -> 72
  in
  List.iter
    (fun fref -> IO.Output.set_width !fref term_width)
    [IO.Output.std_out;IO.Output.std_err;IO.Output.std_dbg]

let welcome_msg =
  let open Interface.Config in
  Printf.sprintf
    "%s %s%s welcomes you.\n\
    \n\
    For a little help, type \"#help.\"\n"
    package_name
    package_version
    (if build="v"^package_version || build="" then ""
     else " (revision " ^ build ^ ")")

let plugins = ref []

let add_plugin_dir,add_plugin =
  let add_plugin_dir dir =
    Interface.Config.(plugin_dirs := dir :: !plugin_dirs)
  and add_plugin (plugin,desc) =
    let open Dynlink in
    let rec aux = function
      | [] ->
          IO.Output.dprintf
            "The plugin %s was not found in the search path." plugin
      | dir::dirs ->
          try
            loadfile (adapt_filename (Filename.concat dir plugin ^ ".cma")) ;
            plugins := desc :: !plugins
          with Error e -> aux dirs
    in
    aux !Interface.Config.plugin_dirs
  in
  add_plugin_dir,add_plugin

let () =
  List.iter add_plugin [
    "tableXmlExport","XML export of proofs" ;
  ]

let version_msg =
  let open Interface.Config in
  Printf.sprintf
    "%s prover %s, Copyright (C) 2005-2015 Slimmer project.\n\
    This is free software, distributed under the GNU General Public License\n\
    version 2.  There is NO WARRANTY, not even SOUNDNESS nor COMPLETENESS.\n\
    Rev. %s (built with OCaml %s on the %s).\n\
    Features (+/-):%s\n\
    Plugins (+/-):%s\n"
    package_name
    package_version
    (if build="" then "unknown" else build)
    ocaml_version
    build_date
    (let get (s1,s2) = [(match s2 with "" -> "\n - " | _ -> "\n + ") ; s1] in
     String.concat "" (List.flatten (List.map get features)))
    (let get s = [("\n + ") ; s] in
     String.concat "" (List.flatten (List.map get !plugins)))

let print_version () =
  IO.Output.printf ~nl:true "%s" version_msg ;
  exit 0

let usage_msg =
  Printf.sprintf
    "Usage: bedwyr [filename | option]*\n"

let batch       = ref false

(* Bedwyr's working directory *)
let () = IO.Files.bwd := Sys.getcwd ()
module InclFiles =
  Set.Make (struct type t = string let compare = compare end)
let inclfiles = ref (InclFiles.empty)

let _ =
  Prover.System.(
    Arg.parse
      (Arg.align
         [ "-I", Arg.Set batch,
             " Do not enter interactive mode" ;
           "-t", Arg.Unit incr_test_limit,
             " Run tests (use as many times as the #include-depth)" ;
           "-T", Arg.Unit remove_test_limit,
             " Run tests (use as many times as the #include-depth)" ;
           "--filter", Arg.Set use_filter,
             "Use tabling with variables" ;
           "-d", Arg.String (fun s -> definitions := s::!definitions),
             "<s> Add definition" ;
           "-e", Arg.String (fun s -> queries := s::!queries),
             "<s> Execute query" ;
           "--freezing", Arg.Set_int Prover.Proofsearch.freezing_point,
             "<n> Enable backward chaining and set its limit" ;
           "--saturation", Arg.Set_int Prover.Proofsearch.saturation_pressure,
             "<n> Enable forward chaining and set its limit" ;
           "--version", Arg.Unit print_version,
             " Display version info and exit" ;
           "-D", Arg.Set IO.Output.debug,
             " Print debugging information" ;
           "--colours", Arg.Int (fun n -> IO.Output.set_colours n),
            "<n> Number of supported colours (256 or 8; 0 do disable)" ;
         ])
      (fun f -> session := f::!session)
      usage_msg ;
    session := List.rev (!session) ;
    definitions := List.rev (!definitions) ;
    queries := List.rev (!queries))

let _ =
  Interface.Run.reload () ;
  Prover.System.Status.exit_if () ;
  List.iter
    (function query -> ignore (Interface.Run.query_string query))
    !Prover.System.queries ;
  if !batch then Prover.System.Status.exit_with ()
  else begin
    IO.Output.printf ~nl:true "%s" welcome_msg ;
    Interface.Run.queries_channel stdin
  end
